### PR TITLE
Automatically upload results into S3 if greater than 4mb

### PR DESCRIPTION
**Pull request by dkolas**

### CODE REVIEW

File: src/steamship/app/lambda_handler.py [hunk 1]
```diff
@@ -159,9 +151,55 @@ def handler(event: Dict, context: Dict = None) -> dict:
...
+                    operation=SignedUrl.Operation.WRITE,
+                )
+            )
+            signed_url = signed_url_response.data.signed_url
+            logging.info(f"Got signed url for writing: {signed_url}")
+
+            upload_to_signed_url(signed_url, json.dumps(result.get("data", None)).encode("UTF-8"))
+
+            # Now remove raw data and replace with bucket
+            result["data"] = None
```
Comment (EniasCailliau) [inline src/steamship/app/lambda_handler.py:195]:
```suggestion
            del result["data"]
```

File: src/steamship/app/lambda_handler.py [hunk 2]
```diff
@@ -159,9 +151,55 @@ def handler(event: Dict, context: Dict = None) -> dict:
...
+            result["dataBucket"] = str(SignedUrl.Bucket.PLUGIN_DATA.value)
```
Comment (EniasCailliau) [inline src/steamship/app/lambda_handler.py:196]:
```suggestion
            result["dataBucket"] = SignedUrl.Bucket.PLUGIN_DATA.value
```

Isn't the value already a str?

File: src/steamship/app/lambda_handler.py [hunk 3]
```diff
@@ -159,9 +151,55 @@ def handler(event: Dict, context: Dict = None) -> dict:
...
+        response = _handler(event, client, context)
+
+        result = response.dict(by_alias=True)
+        # When created with data > 4MB, data is uploaded to a bucket.
+        # This is a very ugly way to get the deep size of this object
+        data_size = sys.getsizeof(json.dumps(result.get("data", {})))
+        logging.info(f"Response data size {data_size}")
+        if data_size > 4e6 and invocation_context.invocable_type == "plugin":
+            logging.info("Response data size >4MB, must upload to bucket")
+            # data_bucket: Optional[SignedUrl.Bucket] = None
```
Comment (EniasCailliau) [inline src/steamship/app/lambda_handler.py:178]:
```suggestion
```

File: src/steamship/app/lambda_handler.py [hunk 4]
```diff
@@ -159,9 +151,55 @@ def handler(event: Dict, context: Dict = None) -> dict:
...
+            # data_filepath: Optional[str] = None
```
Comment (EniasCailliau) [inline src/steamship/app/lambda_handler.py:179]:
```suggestion
```

File: src/steamship/app/lambda_handler.py [hunk 5]
```diff
@@ -159,9 +151,55 @@ def handler(event: Dict, context: Dict = None) -> dict:
...
+
+            filepath = str(uuid.uuid4())
+            signed_url_response = client.get_space().create_signed_url(
+                SignedUrl.Request(
+                    bucket=SignedUrl.Bucket.PLUGIN_DATA,
+                    filepath=filepath,
+                    operation=SignedUrl.Operation.WRITE,
+                )
+            )
```
Comment (EniasCailliau) [inline src/steamship/app/lambda_handler.py:188]:
```suggestion
            ).data.signed_url
```

File: src/steamship/app/lambda_handler.py [hunk 6]
```diff
@@ -159,9 +151,55 @@ def handler(event: Dict, context: Dict = None) -> dict:
...
+        # This is a very ugly way to get the deep size of this object
+        data_size = sys.getsizeof(json.dumps(result.get("data", {})))
+        logging.info(f"Response data size {data_size}")
+        if data_size > 4e6 and invocation_context.invocable_type == "plugin":
+            logging.info("Response data size >4MB, must upload to bucket")
+            # data_bucket: Optional[SignedUrl.Bucket] = None
+            # data_filepath: Optional[str] = None
+
+            filepath = str(uuid.uuid4())
+            signed_url_response = client.get_space().create_signed_url(
```
Comment (EniasCailliau) [inline src/steamship/app/lambda_handler.py:182]:
```suggestion
            signed_url = client.get_space().create_signed_url(
```

File: src/steamship/app/lambda_handler.py [hunk 7]
```diff
@@ -159,9 +151,55 @@ def handler(event: Dict, context: Dict = None) -> dict:
...
+                message="Plugin/App handler was unable to create Steamship client.",
+                exception=ex,
+            ).dict(by_alias=True)
+
+        response = _handler(event, client, context)
+
+        result = response.dict(by_alias=True)
+        # When created with data > 4MB, data is uploaded to a bucket.
+        # This is a very ugly way to get the deep size of this object
+        data_size = sys.getsizeof(json.dumps(result.get("data", {})))
```
Comment (EniasCailliau) [inline src/steamship/app/lambda_handler.py:174]:
```suggestion
        data = json.dumps(result.get("data", None)).encode("UTF-8")
        data_size = sys.getsizeof(data)
```

1. Cache the json.dumps result for upload
2. Get the size of the utf-8 encoded version, as its size is different from the non-encoded version.

File: src/steamship/app/lambda_handler.py [hunk 8]
```diff
@@ -159,9 +151,55 @@ def handler(event: Dict, context: Dict = None) -> dict:
...
+                SignedUrl.Request(
+                    bucket=SignedUrl.Bucket.PLUGIN_DATA,
+                    filepath=filepath,
+                    operation=SignedUrl.Operation.WRITE,
+                )
+            )
+            signed_url = signed_url_response.data.signed_url
+            logging.info(f"Got signed url for writing: {signed_url}")
+
+            upload_to_signed_url(signed_url, json.dumps(result.get("data", None)).encode("UTF-8"))
```
Comment (EniasCailliau) [inline src/steamship/app/lambda_handler.py:192]:
```suggestion
            upload_to_signed_url(signed_url, data)
```

File: tests/assets/plugins/importers/plugin_file_importer_by_url.py [hunk 9]
```diff
@@ -29,23 +25,8 @@ def config_cls(self) -> Type[Config]:
...
-        logging.info(f"Got signed url for writing: {signed_url}")
-        bytes = TEST_DOC.encode("utf-8")
-        upload_to_signed_url(signed_url, bytes)
-        return Response(
-            data=RawDataPluginOutput(
-                bucket=SignedUrl.Bucket.PLUGIN_DATA, filepath=filepath, mime_type=MimeTypes.MKD
-            )
-        )
+        bytes = (100000 * TEST_DOC).encode("utf-8")
+        return Response(data=RawDataPluginOutput(_bytes=bytes, mime_type=MimeTypes.MKD))
```
Comment (EniasCailliau) [inline tests/assets/plugins/importers/plugin_file_importer_by_url.py:29]: Looking clean!

**Review by EniasCailliau — APPROVED** [review]
Only 2 code files were touched, yet a significant impact! 

Added some code suggestions for you to look at. Happy to get this merged in as soon as the engine is merged.

File: src/steamship/app/lambda_handler.py (hunk 2 above)
Comment (dkolas) [reply src/steamship/app/lambda_handler.py:196]: Gah, yes :)

**Review by EniasCailliau — APPROVED** [review]
LGTM! 👍

**Pull request merged**